File: UI/src/app/util/calculations.ts
import * as _ from 'lodash';

import { PlayerDeckState, Rarity, CollectionCardState } from '../missing-cards/missing-cards-page';

const wcCommonDropRate: number = 1 / 3;         // 0.33
const wcUncommonDropRate: number = 1 / 5;       // 0.20
const wcRareDropRate: number = 1 / 10;          // 0.10, not accurate according to real drop rates
const wcMythicRareDropRate: number = 1 / 24;    // 0.04

// sum of (progress with card * deck total worth) per deck where that card is missing
function calcCardWorth(collectionCard: CollectionCardState, playerDecks: PlayerDeckState[]): number {
    // get decks containing that card
    const containingDecks = playerDecks.filter(d => _.some(d.cards, (c: CollectionCardState) => collectionCard.mtgaId === c.mtgaId));

    // get decks that miss that card
    let cardWorth = 0;
    for (let i = 0; i < containingDecks.length; i++) {
        const containingDeck = containingDecks[i];

        const ownedCount = collectionCard.ownedCount;
        const requiredCount = containingDeck.cards.find(c => c.mtgaId === collectionCard.mtgaId).requiredCount;

        if (requiredCount < ownedCount) {
            const deckProgressWithCard = ((containingDeck.totalOwnedCards + 1) / containingDeck.totalDeckCards) * 100;
            const deckTotalWorth = calcDeckTotalWorth(containingDeck);

            cardWorth += (deckProgressWithCard * deckTotalWorth);
        }
    }

    return cardWorth;
}

function calcDeckTotalWorth(deckState: PlayerDeckState): number {
    const nrOfCards = deckState.cards.length;

    const weightCommon = deckState.cards.filter(c => c.rarity === Rarity.Common).length * wcCommonDropRate;
    const weightUncommon = deckState.cards.filter(c => c.rarity === Rarity.Uncommon).length * wcUncommonDropRate;
    const weightRare = deckState.cards.filter(c => c.rarity === Rarity.Rare).length * wcRareDropRate;
    const weightMythicRare = deckState.cards.filter(c => c.rarity === Rarity['Mythic Rare']).length * wcMythicRareDropRate;
    const totalWeight = weightCommon + weightUncommon + weightRare + weightMythicRare;

    const deckWorth = nrOfCards - totalWeight;

    return deckWorth;
}

function calcDeckProgressInPercent(deckState: PlayerDeckState): number {
    return 
}

File: MtgaDeckBuilder.Api/Layout/Calculations.cs
using MtgaDeckBuilder.Api.Controllers.Dtos;
using MtgaDeckBuilder.Api.Model;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace MtgaDeckBuilder.Api.Layout
{
    public class Calculations
    {
        private static readonly float WcCommonDropRate = 1 / 3;         // 0.33
        private static readonly float WcUncommonDropRate = 1 / 5;       // 0.20
        private static readonly float WcRareDropRate = 1 / 10;          // 0.10, not representing real drop rate
        private static readonly float WcMythicRareDropRate = 1 / 24;    // 0.04

        public static int CalculateWildcardWorthiness(CollectionCardDto collectionCardDto, IEnumerable<PlayerDeckDto> playerDeckDtos)
        {
            // get decks containing the card
            var containingDecks = playerDeckDtos
                .Where(pd => pd.Cards.Any(dc => dc.MtgaId == collectionCardDto.MtgaId))
                .ToArray();

            // get decks that miss that card
            var wildcardWorhiness = 0;

            for (var i = 0; i < containingDecks.Length; i++)
            {
                var containingDeck = containingDecks[i];

                var ownedCount = collectionCardDto.OwnedCount;
                var requiredCount = containingDeck.Cards.Single(c => c.MtgaId == collectionCardDto.MtgaId).RequiredCount;

                if (ownedCount < requiredCount)
                {
                    var deckCompletenessWithCard = ((containingDeck.TotalOwnedDeckCards + 1) * 100) / containingDeck.TotalDeckCards;

                    wildcardWorhiness += (deckCompletenessWithCard * containingDeck.Worth);
                }
            }

            return wildcardWorhiness;
        }

        public static float CalculateDeckWorth(PlayerDeckDto playerDeckDto, IEnumerable<CollectionCardDto> collectionCards)
        {
            var nrOfCards = playerDeckDto.Cards.Count();
            var deckWorth = 0f;

            foreach (var deckCard in playerDeckDto.Cards)
            {
                var collectionCard = collectionCards.Single(cc => cc.MtgaId == deckCard.MtgaId);
                var dropRate = GetRarityDropRate(collectionCard.Data.Rarity);

                deckWorth += (deckCard.RequiredCount * dropRate);
            }

            return deckWorth;
        }

        private static float GetRarityDropRate(Rarity rarity)
        {
            switch (rarity)
            {
                case Rarity.Common: return WcCommonDropRate;
                case Rarity.Uncommon: return WcUncommonDropRate;
                case Rarity.Rare: return WcRareDropRate;
                case Rarity.MythicRare: return WcMythicRareDropRate;
                default: return 0;
            }
        }
    }
}

